app/models/unit.rb

```diff
@@ -12,26 +12,39 @@ class Unit < ApplicationRecord
 
   # validations
   validates :name, :yoksis_id, :status, :type, :instruction_type,
-            presence: true, strict: true
+            presence: true
   validates :yoksis_id,
-            uniqueness: true, strict: true
+            uniqueness: true
   validates :name,
             uniqueness: { scope: %i[ancestry status] }
 
   # callbacks
   before_validation do
-    self.name = name.capitalize_all
+    self.name = name.capitalize_all if name
   end
 
-  # STI helpers
+  # list all unit types
   def self.types
     descendants.map(&:name)
   end
 
-  # Dynamically generate scopes for STI-related models
-  def self.generate_scopes
-    types.each do |type|
-      scope type.tableize.to_sym, -> { where(type: type) }
-    end
-  end
+  # scopes
+  scope :academies, -> { where(type: 'Academy') }
+  scope :art_disciplines, -> { where(type: 'ArtDiscipline') }
+  scope :departments, -> { where(type: 'Department') }
+  scope :disciplines, -> { where(type: 'Discipline') }
+  scope :doctoral_programs, -> { where(type: 'DoctoralProgram') }
+  scope :faculties, -> { where(type: 'Faculty') }
+  scope :institutes, -> { where(type: 'Institute') }
+  scope :interdisciplinary_disciplines, -> { where(type: 'InterdisciplinaryDiscipline') }
+  scope :interdisciplinary_doctoral_programs, -> { where(type: 'InterdisciplinaryDoctoralProgram') }
+  scope :interdisciplinary_master_programs, -> { where(type: 'InterdisciplinaryMasterProgram') }
+  scope :master_programs, -> { where(type: 'MasterProgram') }
+  scope :proficiency_in_art_programs, -> { where(type: 'ProficiencyInArtProgram') }
+  scope :rectorships, -> { where(type: 'Rectorship') }
+  scope :research_centers, -> { where(type: 'ResearchCenter') }
+  scope :science_disciplines, -> { where(type: 'ScienceDiscipline') }
+  scope :undergraduate_programs, -> { where(type: 'UndergraduateProgram') }
+  scope :universities, -> { where(type: 'University') }
+  scope :vocational_schools, -> { where(type: 'VocationalSchool') }
```

Comment:
@serhat bunları eski haline almanın sebebi scope'ların load/initial edilmesiyle ilgili bir sorun muydu?

Reply:
@ecmelkytz çok güzel bir soru, özet olarak cevabı teoride çok güzel görünen Rails'te scope'ları generate etme seçeneğinin, pratikte kötü sonuçlara yol açtığını fark etmiş olmam. Ancak sebep load edilmesiyle ilgili değil, sorunsuz load ediliyordu eski haliyle.

Temelde iki soruna yol açtı, bunlardan ilki ve en önemlisi generate edilen scope'ların method chaining'e izin vermemesiydi. Zaten scope'un tüm esprisi bu olduğu için, bu büyük bir sorundu.

Diğer sebep generate edilen scope'ların sadece ActiveRecord nesneleri için çalışıyor olmasıydı. Yani instance variable'lar üzerinde beklediğim gibi çalışmadı. Oysa ki gelenelsel yolla scope tanımladığımız da hem class düzeyinde, hemde instance düzeyince tıkır tıkır çalışıyor.

Reply:
Dediğin sorunları da çözerek bunu otomatik üretebiliriz ama bence doğrusu bu hali.  Öbür türlü fazla akıllı bir kod oluyor (şu yaştaki deneyimimle artık doğru bulmadığım bir durum).